src/main/java/hudson/plugins/s3/callable/S3GzipCallable.java

```diff
@@ -21,23 +24,47 @@ public S3GzipCallable(String accessKey, Secret secretKey, boolean useRole, Desti
     @Override
     public String invoke(FilePath file) throws IOException, InterruptedException {
         final File localFile = File.createTempFile("s3plugin", ".bin");
+        Upload upload = null;
 
-        try (InputStream inputStream = file.read()) {
-            try (OutputStream outputStream = new FileOutputStream(localFile)) {
-                try (OutputStream gzipStream = new GZIPOutputStream(outputStream, true)) {
-                    IOUtils.copy(inputStream, gzipStream);
-                    gzipStream.flush();
+        try {
+            try (InputStream inputStream = file.read()) {
+                try (OutputStream outputStream = new FileOutputStream(localFile)) {
+                    try (OutputStream gzipStream = new GZIPOutputStream(outputStream, true)) {
+                        IOUtils.copy(inputStream, gzipStream);
+                        gzipStream.flush();
+                    }
                 }
             }
-        }
 
-        final InputStream gzipedStream = new FileInputStream(localFile);
-        final ObjectMetadata metadata = buildMetadata(file);
-        metadata.setContentEncoding("gzip");
-        metadata.setContentLength(localFile.length());
+            final InputStream gzipedStream = new FileInputStream(localFile);
+            final ObjectMetadata metadata = buildMetadata(file);
+            metadata.setContentEncoding("gzip");
+            metadata.setContentLength(localFile.length());
+
+            upload = Uploads.getInstance().startUploading(getTransferManager(), file, gzipedStream, getDest().bucketName, getDest().objectName, metadata);
 
-        Uploads.getInstance().startUploading(getTransferManager(), file, gzipedStream, getDest().bucketName, getDest().objectName, metadata);
+            String md5 = MD5.generateFromFile(localFile);
 
-        return MD5.generateFromFile(localFile);
+            upload.addProgressListener(new ProgressListener() {
+                    @Override
+                    public void progressChanged(ProgressEvent event) {
+                        switch (event.getEventType()) {
+                        case TRANSFER_CANCELED_EVENT:
```

Comment:
To make it work one stuff is needed. Please, put `Upload.abort` to `cleanUp` method of `Uploads` if upload was not done yet. Right now plugin uploads any way even if everything else was aborted.